app/code/Magento/InventoryElasticsearch/Plugin/CatalogSearch/Model/Indexer/Fulltext/Action/DataProvider/StockedProductFilterByInventoryStock.php

```diff
@@ -0,0 +1,143 @@
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryElasticsearch\Plugin\CatalogSearch\Model\Indexer\Fulltext\Action\DataProvider;
+
+use Magento\CatalogInventory\Api\Data\StockStatusInterface;
+use Magento\CatalogInventory\Api\StockConfigurationInterface;
+use Magento\CatalogSearch\Model\Indexer\Fulltext\Action\DataProvider;
+use Magento\Elasticsearch\Model\Config;
+use Magento\Framework\App\ResourceConnection;
+use Magento\InventoryIndexer\Model\StockIndexTableNameResolverInterface;
+use Magento\InventorySalesApi\Model\StockByWebsiteIdResolverInterface;
+
+class StockedProductFilterByInventoryStock
+{
+    /**
+     * @var Config
+     */
+    private $config;
+
+    /**
+     * @var StockConfigurationInterface
+     */
+    private $stockConfiguration;
+
+    /**
+     * @var StockIndexTableNameResolverInterface
+     */
+    private $stockIndexTableNameResolver;
+
+    /**
+     * @var ResourceConnection
+     */
+    private $resourceConnection;
+
+    /**
+     * @var StockByWebsiteIdResolverInterface
+     */
+    private $stockByWebsiteIdResolver;
+
+    /**
+     * @param Config $config
+     * @param StockConfigurationInterface $stockConfiguration
+     * @param StockIndexTableNameResolverInterface $stockIndexTableNameResolver
+     * @param ResourceConnection $resourceConnection
+     * @param StockByWebsiteIdResolverInterface $stockByWebsiteIdResolver
+     */
+    public function __construct(
+        Config $config,
+        StockConfigurationInterface $stockConfiguration,
+        StockIndexTableNameResolverInterface $stockIndexTableNameResolver,
+        ResourceConnection $resourceConnection,
+        StockByWebsiteIdResolverInterface $stockByWebsiteIdResolver
+    ) {
+        $this->config = $config;
+        $this->stockConfiguration = $stockConfiguration;
+        $this->stockIndexTableNameResolver = $stockIndexTableNameResolver;
+        $this->resourceConnection = $resourceConnection;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+    }
+
+    /**
+     * Filter out of stock options for configurable product.
+     *
+     * @param DataProvider $dataProvider
+     * @param array $indexData
+     * @param array $productData
+     * @param int $storeId
+     * @return array
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforePrepareProductIndex(
+        DataProvider $dataProvider,
+        array $indexData,
+        array $productData,
+        int $storeId
+    ): array {
+        if ($this->config->isElasticsearchEnabled() && !$this->stockConfiguration->isShowOutOfStock($storeId)) {
+            $productIds = array_keys($indexData);
+            $stockStatuses = [];
+
+            foreach ($this->getProductIdsByWebsiteIds($productIds) as $websiteId => $productIds) {
+                $stock = $this->stockByWebsiteIdResolver->execute($websiteId);
+                $stockTable = $this->stockIndexTableNameResolver->execute((int)$stock->getStockId());
```

Comment:
looks like we have to provide conditional logic for Default Stock similar to this PR 
https://github.com/magento-engcom/msi/pull/1937

to prevent performance degradation on Default Stock